repos/system_upgrade/el7toel8/actors/checkpcidrivers/tests/unit_test_checkpcidrivers.py

```diff
@@ -0,0 +1,282 @@
+            1,
+        ),
+        (
+            "acard-ahci",
+            "good pci id",
+            None,
+            0,
+        ),
+        (
+            "good driver name",
+            "0x1000:0x0060",
+            True,
+            1,
+        ),
+        (
+            "good driver name",
+            "0x1000:0x0064",
+            None,
+            1,
+        ),
+        (
+            "good driver name",
+            "0x1000:0x0065",
+            None,
+            0,
+        ),
+    ],
+)
+def test_basic_checkpcidrivers(
+    monkeypatch,
+    driver_name,
+    pci_id,
+    inhibit_upgrade,
+    num_msgs,
+):
+    """Check the main actor function."""
+    pci_devices.devices[0].driver = driver_name
+    pci_devices.devices[0].pci_id = pci_id
+    monkeypatch.setattr(
+        api,
+        "current_actor",
+        CurrentActorMocked(msgs=[pci_devices, restricted_pci_devices]),
+    )
+    monkeypatch.setattr(
+        reporting,
+        "create_report",
+        create_report_mocked(),
+    )
+    monkeypatch.setattr(
+        CurrentActorMocked,
+        "produce",
+        produce_mocked(),
+    )
+    checkpcidrivers_main()
+    if inhibit_upgrade:
+        assert len(CurrentActorMocked.produce.model_instances) == 1
+        assert CurrentActorMocked.produce.model_instances[0].report[
+            "flags"
+        ] == ["inhibitor"]
+    if num_msgs and not inhibit_upgrade:
+        assert (
+            "flags" not in CurrentActorMocked.produce.model_instances[0].report
+        )
+    assert len(CurrentActorMocked.produce.model_instances) == num_msgs
+
+
+class M(Model):
+    """A model just for the purpose of further test."""
+
+    topic = SystemInfoTopic
+
+    a = fields.Integer(default=0)
+    b = fields.Integer(default=0)
+
+
+@pytest.mark.parametrize(
+    ("driver_names", "key", "exp", "exception"),
+    [
+        # Normal case
+        (
+            [M(a=1, b=2), M(a=1, b=3)],
+            "b",
+            {2: M(a=1, b=2), 3: M(a=1, b=3)},
+            None,
+        ),
+        # Not existing attribute
+        (
+            [M(a=1, b=2), M(a=1, b=3)],
+            "c",
+            None,
+            AttributeError,
+        ),
+        # Not existing attribute, empty Ms
+        (
+            [M(), M()],
+            "c",
+            None,
+            AttributeError,
+        ),
+        # Duplicated attribute values
+        (
+            [M(a=1, b=2), M(a=1, b=2)],
+            "b",
+            None,
+            ValueError,
+        ),
+    ],
+)
+def test_create_dict_lookup_fn(driver_names, key, exp, exception):
+    if exception:
+        with pytest.raises(exception):
+            create_dict_lookup(driver_names, key)
+    else:
+        assert create_dict_lookup(driver_names, key) == exp
+
+
+def test_checkpcidrivers_no_restricted_devs(current_actor_context):
+    """Test if actor works well when no messages provided."""
+    # TODO why it doesn't raises the StopActorExecutionError?
```

Comment:
@vinzenz @pirat89 @fernflower can you help me to understand why there is no error in this test. It should raise StpActorExecutionError from the code.